src/solvers/algebraic_solver.hpp

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2019-2020, Lawrence Livermore National Security, LLC and
+// other Serac Project Developers. See the top-level LICENSE file for
+// details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/**
+ * @file algebraic_solver.hpp
+ *
+ * @brief This file contains the declaration of an iterative solver wrapper
+ */
+
+#ifndef ALGEBRAIC_SOLVER
+#define ALGEBRAIC_SOLVER
+
+#include <memory>
+#include <optional>
+#include <variant>
+
+#include "common/serac_types.hpp"
+#include "mfem.hpp"
+
+namespace serac {
+
+/**
+ * Wraps a (currently iterative) system solver and handles the configuration of linear
+ * or nonlinear solvers
+ */
+class AlgebraicSolver : public mfem::Solver {
+public:
+  // TODO: Eliminate this once a dependency injection approach is used for the solvers
```

Comment:
What do you mean by this TODO? What is the proposed dependency injection approach?

Reply:
That refers to the syntax I proposed [here](https://github.com/LLNL/serac/issues/118#issuecomment-668805621), in which a fully-constructed `AlgebraicSolver` is passed to a solver constructor, i.e.
```cpp
AlgebraicSolver as(MPI_COMM_WORLD, lin_params, nonlin_params);
NonlinearSolidSolver solver(as, /* other params */);
// ready to use
```

instead of 
```cpp
NonlinearSolidSolver solver(/* stuff */);
// invalid state
solver.set_params(lin_params, nonlin_params);
solver.completeSetup();
// now valid and ready to use
```